### ISSUE TITLE
keepassxc: Auto-Type feature not available

### ISSUE THREAD

**Issue by mythsunwind** [NixOS/nixpkgs#94715]
**Describe the bug**
The feature to auto-type in a password is not working on 2.6.0 version of keepassxc on macOS. I also saw the issue with the a former version 2.5.x.

**To Reproduce**
1. Open or create a db
2. Add new entry with user and password and enable Auto-Type for this entry
3. Select new entry in list
4. Check Auto-Type tab or right click and search for Auto-Type menu item

Auto-Type tab is greyed out and disabled. No menu item.

**Expected behavior**
There should be a menu item for auto-type on the entry and also an Auto-Type tab on the bottom of keepassxc. None of them exist, even though the extension is enabled according to the debug info (see below).

**Screenshots**
<img width="250" alt="Screenshot 2020-08-05 at 17 26 00" src="https://user-images.githubusercontent.com/1278296/89432475-77254a00-d741-11ea-9ae0-3f763f8d24fa.png">

**More info**
cmake contains flag to build auto-type feature with `-DWITH_XC_AUTOTYPE=ON`. I can attach full build log if needed.

**Keepassxc debug info**
KeePassXC - Version 2.6.0

Qt 5.12.7
Debugging mode is disabled.

Operating system: macOS 10.15
CPU architecture: x86_64
Kernel: darwin 19.5.0

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- TouchID

Cryptographic libraries:
 libgcrypt 1.8.6

**Notify maintainers**
@jonafato @turion 

**Metadata**
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.5.0, macOS 10.15.5`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.6`
 - channels(sven): `"nixpkgs-20.09pre237079.d418434d127"`
 - channels(root): `"nixpkgs-20.09pre228165.00df2371122"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by flokli** [comment]
cc @NixOS/darwin-maintainers

**Comment by turion** [comment]
Can you somehow rule out an upstream bug?

**Comment by mythsunwind** [comment]
> Can you somehow rule out an upstream bug?

Yes, I would say it is not an upstream bug because it did not work in 2.5.4 for me and I found a bug ticket on upstream which shows that selecting auto-type from the right-click menu worked for 2.5.4 on macOS Catalina (the macOS version i use):

https://github.com/keepassxreboot/keepassxc/issues/4586

**Comment by mythsunwind** [comment]
I checked again and the issue still exists on 2.6.1 in the nix package but upstream version of 2.6.1 is fine.

Additionally the nix version of keepassxc 2.6.1 renders parts of the Qt UI with black icons on a black background now :(

**Comment by turion** [comment]
@mythsunwind 

I cannot reproduce this on my machine. Auto-type works for me. My system:

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.77, NixOS, 21.03pre252083.2deeb58f494 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.8`
 - channels(root): `"musnix, nixos-21.03pre268206.536fe36e23a, nixos-hardware"`
 - channels(turion): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Possibly you could try to update your system and see whether bug persists? Also, what window manager do you use? Are you in KDE?

I believe that this is not a specific bug in keepassxc, but rather in the nixpkgs-Qt story in general. There have been bugs in the past with Qt introducing state in home directories and whatnot. Perhaps @ttuegel can share his wisdom what to look out for?

**Comment by ljubitje** [comment]
This bug is ongoing with KeePassXC 2.7.9 on KDE Plasma 6.2.2.
No Auto-Type settings tab is shown and autotype does not work.
I think it is because of some Wayland limitation.

**I have found and tested 3 workarounds, as per [KeePassXC docs](https://github.com/keepassxreboot/keepassxc/blob/develop/docs/topics/AutoType.adoc).**
They all force the app to relay on X11 instead of Wayland.

Solution 1:  Start KeePassXC from terminal with `-platform xcb` flag.
Full command is simply: `keepassxc -platform xcb`

Solution 2: Install KeePassXC as a Flatpak. (How to enable Flatpak is at the bottom of this post.)
Open (KDE Plasma) System Settings ->  Application Permissions -> Flatpak Permissions -> KeePassXC -> Advanced Permissions -> Sockets
Disable (untick) Wayland windowing system.
Click Apply and restart KeePassXC.

Solution 3: Install KeePassXC as a Flatpak. (How to enable Flatpak is at the bottom of this post.)
Open (KDE Plasma) System Settings ->  Application Permissions -> Flatpak Permissions -> KeePassXC -> Advanced Permissions -> (on bottom) Environment -> Add New
Set Name exactly: `QT_QPA_PLATFORM`
Set Value exactly: `xcb`
Click Apply and restart KeePassXC.

Flatpak workarounds work on other distros also! I tested on Debian Bookworm.
If not on KDE Plasma, download and use Flatseal to set advanced Flatpak permissions on your system.

![KeePassXCWaylandWorkaroundSolutionFixFlatpak](https://github.com/user-attachments/assets/eb692e88-6290-4237-9828-98992cb2378b)

How to enable Flatpak on NixOS:
Simplest and reproducible way is to configure flake.nix as per https://github.com/gmodena/nix-flatpak:
```
{
  inputs = {
    # ...
    nix-flatpak.url = "github:gmodena/nix-flatpak"; # unstable branch. Use github:gmodena/nix-flatpak/?ref=<tag> to pin releases.
  };
  outputs = { nix-flatpak, ... }: {
    nixosConfigurations.<host> = nixpkgs.lib.nixosSystem {
      modules = [
        nix-flatpak.nixosModules.nix-flatpak
        ./configuration.nix
      ];
    };
  };
}
```
Then add this to your configuration.nix:
```
  # FLATPAK
  services.flatpak.enable = true;
  services.flatpak.update.onActivation = true; # updates flatpaks at nixos-rebuild
  services.flatpak.update.auto.enable = true;
  services.flatpak.update.auto.onCalendar = "7d";
  # PACKAGES
  services.flatpak.packages = [
    "org.keepassxc.KeePassXC"
    # here you can declare other flatpaks to install
  ];
```
Run nixos-rebuild switch, and you're done!

Extra: If you want a GUI package store, search for kde.Packages.discover and nix-software-center.

**Comment by ljubitje** [comment]
No matter which workaround the user picks, it does not make a reproducible system because it requires manual intervention after installation.

Can you please attempt to enable auto-type feature by setting cmake build flag `-DWITH_XC_AUTOTYPE=ON` as @mythsunwind suggested, if that is still a valid flag of course?
Or consider finding a solution with `QT_QPA_PLATFORM=xcb` or `platform -xcb`.

I am also interested in any other easily-reproducible solutions!

I hope I was of help and a fix will be implemented.
Thanks and all best, Klemen

**Comment by KucharczykL** [comment]
Apparently they are working on Wayland solution: https://github.com/keepassxreboot/keepassxc/issues/2281

**Comment by ljubitje** [comment]
This issue has not been fixed yet.
I'm tagging maintainers in hopes of moving this to completion.
@SigmaSquadron & @BlankParticle

I can also help with this, but I have no prior packaging experience.

**Comment by SigmaSquadron** [comment]
Uh, I need a refresher on this issue. I've only been a maintainer of KeePassXC for a few months.

To be clear:
- Does this work on X11?
- The issue says Wayland is still unsupported. Is this still the case?

**Comment by ljubitje** [comment]
Autotype seems to be working on X11. I have not tested it, but I keep reading that is why people stay on X11.

> That is why I still run an X11 session with keepass2 after finding out, that autotype does not work with wayland.
(https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-2960256672)

Also the workarounds I found seem to suggest the X11 is fine.

As for the official Wayland support, no, it is not supported yet.
The related PR is still pending. https://github.com/keepassxreboot/keepassxc/pull/10905 #

**Comment by SigmaSquadron** [comment]
Then there's nothing we can do right now.

**Comment by ljubitje** [comment]
I'm not sure I understand it enough, but what about compiling with -platform xcb flag?
